From ef31bb6b3c993e16d74830eeabba8c53d04807db Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Wed, 30 Mar 2022 16:39:58 +0300
Subject: [PATCH] [SW-81331] greco: preboot: clear flr reset

If FLR interrupt is received while FLR is handled by FW, FW issues
FLR reset by setting pcie2psoc bit PCIE_AUX_FLR_CTRL register. However
this bit is not self clear. This patch adds support to clear pcie2psoc bit
as part of preboot init sequence.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I2a312e5deb5edac3484017685968947cbacc8c5c
---

diff --git a/arch/greco/common/common.c b/arch/greco/common/common.c
index 6987299..047d93b 100644
--- a/arch/greco/common/common.c
+++ b/arch/greco/common/common.c
@@ -45,6 +45,8 @@
 #include <efuse.h>
 #include <hl.h>
 #include <greco/greco_boot_info.h>
+#include <pcie_aux_regs.h>
+#include <pcie_aux_masks.h>
 
 #define mmPCIE_WRAP_VUART_TX_DATA       0xC15F60
 #define mmPCIE_WRAP_VUART_TX_STATUS     0xC15F64
@@ -647,6 +649,25 @@
 	regmap_writel(0x1, rst_offset);
 }
 
+static void clr_flr_reset(void)
+{
+	u32 val;
+
+	/*
+	 * If FLR reset is issued by FW, the reset
+	 * needs to be cleared by FW as the bit is not
+	 * self clear.
+	 */
+	val = regmap_readl(mmPCIE_AUX_FLR_CTRL);
+	val &= ~(PCIE_AUX_FLR_CTRL_FSM_PCIE2PSOC_RST_SET_MASK);
+	regmap_writel(val, mmPCIE_AUX_FLR_CTRL);
+}
+
+void preboot_reset_init(void)
+{
+	clr_flr_reset();
+}
+
 void axi_enable_drain(void)
 {
 	u32 val;
